### ISSUE TITLE
[Bug] PEP 593-compliant type hints subscripted by `object` (i.e., `typing(|_extensions).Annotated[object, {whatevah}]`) unconditionally ignored

### ISSUE THREAD

**Issue by wesselb** [beartype/beartype#290]
Hey @leycec, CC @hodgespodge

Here we are again with another obscure issue. :) I'm not sure if this is in or out of scope, so please feel free to say so if it is out of scope, but I though I'd open an issue anyway to get your thoughts.

`Annotated` was added in Python 3.9, but `typing_extensions` offers a substitute for Python 3.8.

Using this substitute, we find the following behaviour:

```python
>>> from typing_extensions import Annotated

>>> from beartype.vale import Is

>>> from beartype.door import TypeHint, is_bearable

>>> is_bearable("hello", Annotated[object, Is[lambda value: hasattr(value, "special_attr")]])
True
```

Is this behaviour to be expected, or could something be going wrong?

**Comment by leycec** [comment]
**Gah!** Superb catch. You're right. You always are. Although @beartype tries really hard to support *all* public `typing_extensions` attributes across *all* supported Python versions, we didn't try hard enough here. The public shame I feel is real.

I'll dig into this tonight. With copious luck, tears, sweat, and buckets of fake Halloween blood, I'll push out a new @beartype 0.16.3 patch release resolving this and a few other pending issues by Friday. I promise nothing but hope to deliver something. *Hype train!* :train2:

**Comment by leycec** [comment]
**Yikes.** This issue is far worse than simply `typing_extensions.Annotated` compliance. I grimace. :grimacing: 

On the bright side, @beartype *is* correctly handling `typing_extensions.Annotated` under Python 3.8. That's good. On the dark side, **@beartype currently ignores *all* beartype validators of the form `typing(|_extensions).Annotated[object, .*]`. Why? Because the root `object` superclass is an ignorable type hint in all other contexts. @beartype incorrectly assumes the root `object` superclass is an ignorable type hint in a `typing(|_extensions).Annotated[...]` context, too.

## tl;dr

@hodgespodge: Subscript `typing_extensions.Annotated[{type}, ...]` by any `{type}` *other* than `object`. @beartype will then begin exhibiting the correct behaviour. In this specific case, consider passing the specific type `str` instead: e.g.,

```python
>>> from typing_extensions import Annotated

>>> from beartype.vale import Is

>>> from beartype.door import TypeHint, is_bearable

                                 # .-- see what we did there? no bad "object". only good "str"!
                                 # |
                                 # v   yes, this is painfully dumb. y, @beartype? y u so dumb?
>>> is_bearable("hello", Annotated[str, Is[lambda value: hasattr(value, "special_attr")]])
False
```

Obviously, this is awful. I'm shocked, appalled, and ultimately unsurprised that you're the first user to uncover this. I promise to resolve this *immediately* and kick out a new @beartype 0.16.3 release. Let's do this for the good of humanity and AI alike, everybody! :face_with_thermometer:

**Comment by leycec** [comment]
**Resolved by 76e609dd484339ae.** I'm willing to pretend this didn't happen if everyone else is. :smiling_face_with_tear:

**Comment by wesselb** [comment]
Holy crap, you solved this bloody quickly!! This is incredibly helpful. Thank you so much, @leycec. :)

And I have to say, the number of special cases and edge cases Beartype has to deal with really is unsettling. 😭 You make users of Beartype very very happy by kindly dealing with this mess for all of us.

**Comment by leycec** [comment]
*Awwwwwwwwww.* @wesselb the best Wessel. :hugs: 

Thanks so much for your wondrously gracious words that I will inevitably fail to live up. For my part, I'm beyond impressed with the *absurdly* high level of academic discourse happening over at Plum. I'm reasonably confident that [you just wrote an arXiv pre-print over here](https://github.com/beartype/plum/pull/119#issuecomment-1756226426) at PR beartype/plum#119. Like, seriously!

> Define varargs as the corresponding infinite set of signatures:

```
(float, *int) = {(float), (float, int), (float, int, int), ...}
```

Mind is already blown.

> To fix #117, we wish to define an order such that `(float, *int) < (float, Number)`.

Mind continues to stay blown.

> For two signatures `S` and `T`, let `S <= T` if (1) `S` and `T` contain an equal number of types and (2) the types in `S` are subtypes of the correspond types in `T`. 

So. We're really doin' this, huh?

> Since `(float, *int)` is defined as a set of signature, we wish to extend `<=` to collections of signatures such that `(float, *int) < (float, Number)` holds. We do this in two steps.

We're really doin' this.

> For a collection of signatures `{S1, S2, ...}` and a single signature `T`, let `{S1, S2, ...} <= T` if _one of the following_ is true: `S1 <= T`, `S2 <= T`, _et cetera_. This models the desired condition.

> For two collections of signatures `{S1, S2, ...}` and `{T1, T2, ...}`, let `{S1, S2, ...} <= {T1, T2, ...}` if _all of the following_ are true: `{S1, S2, ...} <= T1`, `{S1, S2, ...} <= T2`, _et cetera_.

> In words, `{S1, S2, ...} <= {T1, T2, ...}` if, for every `Ti`, `Sj <= Ti` for some `Sj`. That is, every element in `{T1, T2, ...}` is preceded by some element in `{S1, S2, ...}`.

> We now investigate whether this is a partial order. We desire a partial order, because then equality means that things are really equal (anti-symmetry).

> It is true that `{S1, S2, ...}  <= {S1, S2, ...}`, so the relation is reflexive.

> Suppose that `{S1, S2, ...} <= {T1, T2, ...} <= {U1, U2, ...}`. For every `Ui`, `Tj <= Ui` for some `Tj`. For that `Tj`, `Sk <= Tj` for that `Sk`. Hence, by transitivity of `<=` for single signatures, for every `Ui`, `Sk <= Ui` for some `Sk`, meaning that `{S1, S2, ...} <= {U1, U2, ...}`.

> For anti-symmetry, we assume that `{S1, S2, ...} <= {T1, T2, ...}` and `{T1, T2, ...} <= {S1, S2, ...}`. If `{T1, T2, ...} = {S1, S2, ...}`, then the relation would be anti-symmetric . Unfortunately, note that `{(int)} == {(int), (Number)}`, because `{(int)} <= {(int), (Number)}` and `{(int), (Number)} <= {(int)}`, so I do not think anti-symmetry holds.

> _However_, if we restrict ourselves to collections of signatures where every two different signatures have a different number of types, then I think anti-symmetry does hold!

> Therefore, the proposed relation is indeed a partial order if we consider sets of signatures a where every two different signatures have a different number of types. Note that this includes (1) singular signatures like `{(S)}` and (2) varargs.

:exploding_head: 

## Speed: It's Pretty Much All Beartype Does

Relatedly, would Plum benefit from space- and time-efficient signature introspection? If efficiency ever becomes a concern there, @beartype can help. I implemented a [*stupidly* micro-optimized `iter_func_args()` generator function for efficiently iterating over callable parameters](https://github.com/beartype/beartype/blob/76e609dd484339aef8221d7105f7f11f8226caff/beartype/_util/func/arg/utilfuncargiter.py#L154) a year or so ago. It's exhaustively unit-tested and extensively leveraged throughout @beartype itself for signature introspection. So, you just know it's solid gold.

I suspect Plum probably just defers to the standard `inspect.signature()` class for introspecting callable signatures, right? That is what the `@beartype` decorator used to do. Then I profiled `@beartype` after somebody justifiably complained and realized (to my mounting horror and shame) that `@beartype` spent most of its time in the `inspect` module rather than in the `beartype` package. That was the birth of `iter_func_args()`, which drank the milkshake of `inspect.signature()` and then chucked the empty husk of that milkshake into the nearest pile of hot steaming compost on our property in Canada.

`inspect`: *I will never trust you again.*

<sup>**P.S.:** What does "mounting horror" even mean? It's probably best not to think about this.</sup>

**Comment by wesselb** [comment]
@leycec, yes, space- and time-efficient signature introspection is most certainly something Plum would benefit from! Your linked `iter_func_args` actually sounds perfect and could basically be used as a drop-in replacement for `inspect.signature`, which we do indeed use.

I've [created an issue here](https://github.com/beartype/plum/issues/121) to write this down. :) 

Does Beartype also happen to implement a blazingly fast function that gets the return type for a pure-Python callable? I suppose this is rather simply done manually, just by `f.__annotations__["return"]`, but I'm just wondering if there are edge cases to think about here.

**Comment by leycec** [comment]
> I've https://github.com/beartype/plum/issues/121 to write this down. :)

Yet again, I see that you win GitHub. :smile: 

> Does Beartype also happen to implement a blazingly fast function that gets the return type for a pure-Python callable?

**Excellent question.** The current answer is that I am lazy, fat, and tired. Therefore...

> I suppose this is rather simply done manually, just by `f.__annotations__["return"]`...

Shamefully, this is the answer. This is what @beartype does.

Currently, there are *no* edge cases in that idiom. There probably will be under a future Python release. It's a good thing that I did nothing to defend us against that inevitability. Oh, well. Let's just play turn-based strategy games until this fails catastrophically like my current Indo-Canadian Pan-Galactic Empire of Prosperity.

**Comment by wesselb** [comment]
> Currently, there are no edge cases in that idiom. There probably will be under a future Python release. It's a good thing that I did nothing to defend us against that inevitability. Oh, well. Let's just play turn-based strategy games until this fails catastrophically like my current Indo-Canadian Pan-Galactic Empire of Prosperity.

Awesome! If you say that there are no edge cases, then that's probably as sure as one can possibly be. :)

Fortunately, [@PhilipVinc's PR](https://github.com/beartype/plum/pull/112) will very soon disentangle signature types and return types, meaning that the current `iter_func_args` can nearly literally just be dropped in. Life's good! :D

**Comment by leycec** [comment]
**Superb!** @PhilipVinc is the physicist we deserve. Oh – and a related heads up that is frightening.

All of this (specifically, the standard `__annotations__` dictionary) is expected to fundamentally change with Python 3.13. Why? Because [PEP 649 – Deferred Evaluation Of Annotations Using Descriptors](https://peps.python.org/pep-0649) goes live for Python 3.13. According to the [official Python 3.12 release notes](https://docs.python.org/3/reference/executionmodel.html#annotation-scopes):

> In Python 3.13, [annotations](https://docs.python.org/3/glossary.html#term-annotation) will also be evaluated in [annotation scopes](https://docs.python.org/3/reference/executionmodel.html#annotation-scopes).

Annotation scopes are *basically* PEP 563 (i.e., `from __future__ import annotations`) on steroids. Under Python < 3.13, the standard `__annotations__` dictionary associated with callables and classes was just a standard dictionary. Under Python ≥ 3.13, each `__annotations__` dictionary will now be **lazily evaluated** (i.e., dynamically computed on the first access of that dictionary from an associated `__annotate__()` dunder method). I guarantee that fails in edge cases we currently support.

Python 3.12 also unleashes fascinating new syntax like:

```python
type ListOrSet[T] = list[T] | set[T]  # <-- "T" remains undefined, but this is fine
type AnimalOrVegetable = Animal | "Vegetable"  # <-- "Animal" has yet to be defined, but this is fine
type RecursiveList[T] = T | list[RecursiveList[T]]  # <-- self-recursive type hints, but this is fine

class Animal[T: str]:  # <-- syntactic sugar for "T = TypeVar('T', covariant=True, bound=str)"
    def method1(self) -> T:
        ...

def func[T](a: T, b: T) -> T:  # <-- more madness
    ...
```

It's all pretty intense and speculative. I'm withholding judgement and wrathful anger until all of this inevitably breaks both Plum and @beartype, at which point @leycec uselessly but vengefully releases the hounds of whinging, hand-wringing, and frowning cat memes on Python's `typing-sig` mailing list.

**tl;dr:** Python annotations just got *real* nasty and about to get even *more* nasty.

**Comment by wesselb** [comment]
Oh, dear god, I was not aware that this is coming... This seems like it is going to give rise to a slew of issues. 😭 I think I'll just put my head back in the sand and forget that I just read this. 🫠
